src/Time/History.hpp

```diff
@@ -0,0 +1,192 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <deque>
+#include <iterator>
+#include <pup.h>
+#include <tuple>
+#include <utility>
+
+#include "Parallel/PupStlCpp11.hpp"
+#include "Time/Time.hpp"
+
+namespace TimeSteppers {
+
+template <typename Vars, typename DerivVars>
+class HistoryIterator;
+
+/// \ingroup TimeSteppersGroup
+/// History data used by a TimeStepper.
+/// \tparam Vars type of variables being integrated
+/// \tparam DerivVars type of derivative variables
+template <typename Vars, typename DerivVars>
+class History {
+ public:
+  using value_type = Time;
+  using const_reference = const value_type&;
+  using const_iterator = HistoryIterator<Vars, DerivVars>;
+  using difference_type =
+      typename std::iterator_traits<const_iterator>::difference_type;
+  using size_type = size_t;
+
+  History() = default;
+  History(const History&) = delete;
+  History(History&&) = default;
+  History& operator=(const History&) = delete;
+  History& operator=(History&&) = default;
+  ~History() = default;
+
+  /// Add a new set of values at the given time.
+  void insert(Time time, Vars value, DerivVars deriv) noexcept;
+
+  /// Add a new set of values at the start of the history.  This is
+  /// often convenient for setting initial data.
+  void insert_initial(Time time, Vars value, DerivVars deriv) noexcept;
+
+  /// Mark all data before the passed point in history as unneeded so
+  /// it can be removed.  Calling this directly should not often be
+  /// necessary, as it is handled internally by the time steppers.
+  void mark_unneeded(const const_iterator& first_needed) noexcept;
+
+  /// Access to past values.  These directly return the Time of the
+  /// relevant entry.  The other data can be accessed through the
+  /// iterators using HistoryIterator::value() and
+  /// HistoryIterator::derivative().
+  //@{
+  const_iterator begin() const noexcept { return data_.begin(); }
```

Comment:
I am confused by the comment; these return `HistoryIterator`s (which when dereferenced give the Time)

Reply:
Yeah, it's poorly worded.  I want some comment to this effect in the `History` documentation because I'd look for it there and not in the iterator documentation, and having no explanation in `History` of how to get values out sounds confusing.  (I should add a similar thing to `BoundaryHistory`.)  I'll improve it.